File: C++/105.cpp
__________________________________________________________________________________________________
12ms
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.size()==0)
            return NULL;
         stack<int> s;
         stack<TreeNode *> st;
         TreeNode *t,*r,*root;
         int i,j,f;

         f=i=j=0;
         s.push(preorder[i]);

         root = new TreeNode(preorder[i]);
         st.push(root);
         t = root;
         i++;

         while(i<preorder.size())
         {
             if(!st.empty() && st.top()->val==inorder[j])
             {
                 t = st.top();
                 st.pop();
                 s.pop();
                 f = 1;
                 j++;
             }
             else
             {
                 if(f==0)
                 {
                     s.push(preorder[i]);
                     t -> left = new TreeNode(preorder[i]);
                     t = t -> left;
                     st.push(t);
                     i++;
                 }
                 else 
                 {
                     f = 0;
                     s.push(preorder[i]);
                     t -> right = new TreeNode(preorder[i]);
                     t = t -> right;
                     st.push(t);
                     i++;
                 }
             }
         }

         return root;
     } 
};
__________________________________________________________________________________________________
14164 kb
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        if(preorder.size()==0){
            return nullptr;
        }
        
        stack<TreeNode*> s;
        TreeNode* p;
        TreeNode* root = new TreeNode(preorder[0]);
        int j=0;
        s.push(root);
        
        for(int i=1;i<preorder.size();i++){
            if(s.top()->val!=inorder[j] && s.top()->left==nullptr){
                TreeNode* node = new TreeNode(preorder[i]);
                s.top()->left=node;
                s.push(node);
            }else{
                while(s.empty()==false && s.top()->val == inorder[j]){
                    p=s.top();
                    s.pop();
                    j++;
                }
                p->right=new TreeNode(preorder[i]);
                s.push(p->right);
            }
        }
        return root;
    }
};
__________________________________________________________________________________________________

File: Java/105.java
__________________________________________________________________________________________________
sample 1 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    int in, pre;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return buildTreeHelper(preorder, inorder, Long.MAX_VALUE);
    }
    
    public TreeNode buildTreeHelper(int[] preorder, int[] inorder, long val) {
        if (in >= inorder.length || val == inorder[in])
            return null;
        
        TreeNode root = new TreeNode(preorder[pre++]);
        root.left = buildTreeHelper(preorder, inorder, root.val);
        in++;
        root.right = buildTreeHelper(preorder, inorder, val);
        return root;
    }
}
__________________________________________________________________________________________________
sample 34448 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0) {
            return null;
        }
        
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode root = new TreeNode(preorder[0]);
        TreeNode prev = root;
        for (int i = 1, j = 0; i < preorder.length; i++) {
            if (prev.val != inorder[j]) {
                prev.left = new TreeNode(preorder[i]);
                stack.push(prev);
                prev = prev.left;
            } else {
                j++;
                while (!stack.isEmpty() && stack.peek().val == inorder[j]) {
                    prev = stack.pop();
                    j++;
                }
                prev = prev.right = new TreeNode(preorder[i]);
            }   
        }
        return root;
    }
}
__________________________________________________________________________________________________
